server/src/main/java/org/elasticsearch/action/get/TransportGetFromTranslogAction.java

```diff
@@ -0,0 +1,182 @@
+
+    private final IndicesService indicesService;
+
+    @Inject
+    public TransportGetFromTranslogAction(TransportService transportService, IndicesService indicesService, ActionFilters actionFilters) {
+        super(NAME, transportService, actionFilters, Request::new, ThreadPool.Names.GET);
+        this.indicesService = indicesService;
+    }
+
+    @Override
+    protected void doExecute(Task task, Request request, ActionListener<Response> listener) {
+        final GetRequest getRequest = request.getRequest();
+        final ShardId shardId = request.shardId();
+        IndexService indexService = indicesService.indexServiceSafe(shardId.getIndex());
+        IndexShard indexShard = indexService.getShard(shardId.id());
+        assert indexShard.routingEntry().isPromotableToPrimary();
+        assert getRequest.realtime();
+        ActionListener.completeWith(listener, () -> {
+            var result = indexShard.getService()
+                .getFromTranslog(
+                    getRequest.id(),
+                    getRequest.storedFields(),
+                    getRequest.realtime(),
+                    getRequest.version(),
+                    getRequest.versionType(),
+                    getRequest.fetchSourceContext(),
+                    getRequest.isForceSyntheticSource()
+                );
+            long segmentGeneration = -1;
+            if (result == null) {
+                Engine engine = indexShard.getEngineOrNull();
+                if (engine == null) {
+                    throw new AlreadyClosedException("engine closed");
+                }
+                segmentGeneration = ((InternalEngine) engine).getLastUnsafeSegmentGenerationForGets();
+            }
+            return new Response(result, segmentGeneration);
+        });
+    }
+
+    public static class Request extends ActionRequest {
+
+        private final GetRequest getRequest;
+        private final ShardId shardId;
+
+        public Request(GetRequest getRequest, ShardId shardId) {
+            this.getRequest = Objects.requireNonNull(getRequest);
+            this.shardId = Objects.requireNonNull(shardId);
+        }
+
+        public Request(StreamInput in) throws IOException {
+            super(in);
+            getRequest = new GetRequest(in);
+            shardId = new ShardId(in);
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            super.writeTo(out);
+            getRequest.writeTo(out);
+            shardId.writeTo(out);
+        }
+
+        public GetRequest getRequest() {
+            return getRequest;
+        }
+
+        public ShardId shardId() {
+            return shardId;
+        }
+
+        @Override
+        public ActionRequestValidationException validate() {
+            return null;
+        }
+
+        @Override
+        public String toString() {
+            return "GetFromTranslogRequest{" + "getRequest=" + getRequest + ", shardId=" + shardId + "}";
```

Comment:
Maybe just
```suggestion
            return "translog " + getRequest;
```

as GetRequest.toString already provides useful information?

Reply:
doesn't seem to include the shardID though.